### PR TITLE
Friesian example recall codes update

### PR DESCRIPTION
**Pull request by devWangBin**
Primary Changes：
1. Add readme file;
2. improve some parameter settings and add spark-submit mode;
3. remove use_spark, driver memory and other useless parameters;
4. Fix some typos；

### CODE REVIEW

File: python/friesian/example/recall/generate_test_data.py [hunk 1]
```diff
@@ -175,8 +176,8 @@ def generate_data(args):
 
     parser.add_argument('--verbose', type=bool, default=True,
                         help='whether print more detail information')
-    parser.add_argument('--use_spark', type=bool, default=True,
-                        help='whether to use spark to generate vector embeddings')
+    parser.add_argument('--use_spark', action='store_true',
+                        help='Use spark to generate vector embeddings')
```
Comment (hkvision) [inline python/friesian/example/recall/generate_test_data.py:180]: store_true will default to be False?

File: python/friesian/example/recall/generate_test_data.py [hunk 2]
```diff
@@ -69,6 +66,10 @@ def gen_vector(x, vec_dim):
 
 
 def generate_data(args):
+    from bigdl.orca import init_orca_context, OrcaContext, stop_orca_context
```
[the remark this answers is not in the document]
Comment (devWangBin) [reply python/friesian/example/recall/generate_test_data.py:69]: Yes, I agree with you. The modified content will be updated later and thank you for your advice.

File: python/friesian/example/recall/generate_test_data.py [hunk 3]
```diff
@@ -175,8 +176,8 @@ def generate_data(args):
 
     parser.add_argument('--verbose', type=bool, default=True,
```
[the remark this answers is not in the document]
Comment (devWangBin) [reply python/friesian/example/recall/generate_test_data.py:177]: Yes, you're right. I'll revise it as you suggest

File: python/friesian/example/recall/generate_test_data.py (hunk 1 above)
Comment (devWangBin) [reply python/friesian/example/recall/generate_test_data.py:180]: yes, it will be False

Comment (hkvision) [reply python/friesian/example/recall/generate_test_data.py:180]: Then why not set default to be True?

Comment (devWangBin) [reply python/friesian/example/recall/generate_test_data.py]: Maybe we should use stroe_false and change the name to not_use_spack, then the default will be True

File: python/friesian/example/recall/generate_test_data.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nyamashi) [reply python/friesian/example/recall/generate_test_data.py]: 👍

File: python/friesian/example/recall/generate_test_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nyamashi) [reply python/friesian/example/recall/generate_test_data.py]: 👍  It looks good!

File: python/friesian/example/recall/search.py [hunk 4]
```diff
@@ -114,7 +109,7 @@ def search(args):
     with StopWatch("do_search spark >>>>>>") as sw:
         df = spark.read.parquet(args.parquet_path)
         print('Total number of items: ', df.count())
-        rdd = df.rdd.repartition(num_executors)  # Each node runs one faiss task
+        rdd = df.rdd.repartition(args.num_repartition)  # Each node runs one faiss task
```
Comment (hkvision) [inline python/friesian/example/recall/search.py:112]: num_partition should be the same as num_executors right?

File: python/friesian/example/recall/search.py [hunk 5]
```diff
@@ -135,14 +130,26 @@ def search(args):
...
+    parser.add_argument('--num_nodes', type=int, default=4,
+                        help='The number of nodes to use in the cluster.')
+    parser.add_argument('--executor_cores', type=int, default=8,
+                        help='The executor core number')
+    parser.add_argument('--executor_memory', type=str, default="50g",
+                        help='The executor memory')
+    parser.add_argument('--driver_cores', type=int, default=4,
+                        help='The driver core number')
+    parser.add_argument('--driver_memory', type=str, default="36g",
+                        help='The driver memory')
```
Comment (hkvision) [inline python/friesian/example/recall/search.py:146]: What about removing these two arguments and just give default values?
driver memory doesn't need that much right?

File: python/friesian/example/recall/README.md [hunk 6]
```diff
@@ -1,6 +1,100 @@
-https://github.com/facebookresearch/faiss/blob/main/INSTALL.md
+# Recall items by using faiss
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:1]: Offline Recall using Faiss on Spark

File: python/friesian/example/recall/README.md [hunk 7]
```diff
@@ -1,6 +1,100 @@
...
+This example demonstrates how to use BigDL Friesian 
+to get the recalled items by using the retrieval algorithms provided by faiss.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:3]: modify this to add the keywords: offline recall, efficiently, spark

File: python/friesian/example/recall/README.md [hunk 8]
```diff
@@ -1,6 +1,100 @@
...
+
+```
+conda create -n bigdl python=3.7  # "bigdl" is the conda environment name, you can use any name you like.
+conda activate bigdl
+conda install -c pytorch faiss-cpu 
+pip install --pre --upgrade bigdl-friesian
+```
+
+## Generate data
+You can generate some test data for testing, example command:
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:17]: Add more description what data will be generated.

File: python/friesian/example/recall/README.md [hunk 9]
```diff
@@ -1,6 +1,100 @@
...
+__Options for generate_test_data:__
+* `row_nums`: The number of vectors to be generated. Default to be 200000.
+* `vec_dim`: The dimension of vector. Default to be 256.
+* `use_spark`: Use spark to generate vector embeddings, saving to parquet file. Default to be False.
+* `verbose`: Print more detail information. Default to be False.
+* `index_type`: The faiss index_type: FlatL2 or IVFFlatL2. Default to be FlatL2.
+* `emb_path`: The path to save vector embeddings. Default to be ./emb_vecs.pkl.
+* `dict_path`: The path to save item_dict. Default to be ./item_dict.pkl.
+* `index_save_path`: The path to save faiss index data. Default to be ./index_FlatL2.pkl.
+* `parquet_path`: The path to save vector embeddings with spark, only work when use_spark is True. Default to be ./data.parquet/.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:86]: Put this into generate data

File: python/friesian/example/recall/README.md [hunk 10]
```diff
@@ -1,6 +1,100 @@
-https://github.com/facebookresearch/faiss/blob/main/INSTALL.md
+# Recall items by using faiss
+This example demonstrates how to use BigDL Friesian 
+to get the recalled items by using the retrieval algorithms provided by faiss.
 
-Need to install via conda to get reasonable performance on CPU:
+## Prepare the environment
+We recommend you to use [Anaconda](https://www.anaconda.com/distribution/#linux) to prepare the environments, especially if you want to run on a yarn cluster (yarn-client mode only).
+Also Need to install faiss via conda to get reasonable performance on CPU, referring [faiss install](https://github.com/facebookresearch/faiss/blob/main/INSTALL.md).
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:7]: Also Need -> Faiss needs to be installed via conda
reasonable -> promising
referring -> . Please refer to ... for more installation guidance.

File: python/friesian/example/recall/README.md (hunk 9 above)
Comment (hkvision) [reply python/friesian/example/recall/README.md:86]: Also need to briefly describe these paths would be the input of search.py?

File: python/friesian/example/recall/README.md [hunk 11]
```diff
@@ -1,6 +1,100 @@
...
+    --parquet_output_path /path/to/the/folder/to/save/retrieval/results 
+```
+
+__Options for generate_test_data:__
+* `row_nums`: The number of vectors to be generated. Default to be 200000.
+* `vec_dim`: The dimension of vector. Default to be 256.
+* `use_spark`: Use spark to generate vector embeddings, saving to parquet file. Default to be False.
+* `verbose`: Print more detail information. Default to be False.
+* `index_type`: The faiss index_type: FlatL2 or IVFFlatL2. Default to be FlatL2.
+* `emb_path`: The path to save vector embeddings. Default to be ./emb_vecs.pkl.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:83]: Need to mention what paths can be HDFS path, also for the paths in search.py

File: python/friesian/example/recall/generate_test_data.py [hunk 12]
```diff
@@ -173,10 +174,10 @@ def generate_data(args):
     parser.add_argument('--header_len', type=int, default=8,
                         help='The length of the unique item header')
 
-    parser.add_argument('--verbose', type=bool, default=True,
-                        help='whether print more detail information')
-    parser.add_argument('--use_spark', type=bool, default=True,
-                        help='whether to use spark to generate vector embeddings')
+    parser.add_argument('--verbose', action='store_true',
+                        help='Print more detail information')
```
Comment (hkvision) [inline python/friesian/example/recall/generate_test_data.py:178]: detail -> detailed

File: python/friesian/example/recall/search.py (hunk 5 above)
Comment (devWangBin) [reply python/friesian/example/recall/search.py:146]: For consistency with the other examples, I think we should keep these parameters.

File: python/friesian/example/recall/README.md (hunk 6 above)
Comment (devWangBin) [reply python/friesian/example/recall/README.md:1]: Okay~

File: python/friesian/example/recall/README.md (hunk 7 above)
Comment (devWangBin) [reply python/friesian/example/recall/README.md:3]: got it

File: python/friesian/example/recall/README.md (hunk 8 above)
Comment (devWangBin) [reply python/friesian/example/recall/README.md:17]: ok

File: python/friesian/example/recall/README.md (hunk 10 above)
Comment (devWangBin) [reply python/friesian/example/recall/README.md:7]: ok~

File: python/friesian/example/recall/README.md (hunk 9 above)
Comment (devWangBin) [reply python/friesian/example/recall/README.md:86]: ok~

File: python/friesian/example/recall/search.py (hunk 4 above)
Comment (devWangBin) [reply python/friesian/example/recall/search.py:112]: done

File: python/friesian/example/recall/search.py (hunk 5 above)
Comment (devWangBin) [reply python/friesian/example/recall/search.py:146]: removed arguments of driver

File: python/friesian/example/recall/README.md [hunk 13]
```diff
@@ -1,6 +1,107 @@
-https://github.com/facebookresearch/faiss/blob/main/INSTALL.md
+# Offline Recall with Faiss on Spark
+This example demonstrates how to use the retrieval algorithms provided by faiss 
+to perform the offline recall task efficiently on Spark. 
 
-Need to install via conda to get reasonable performance on CPU:
+
+## Prepare the environment
+We recommend you to use [Anaconda](https://www.anaconda.com/distribution/#linux) to prepare the environments, especially if you want to run on a yarn cluster (yarn-client mode only).
+Faiss needs to be installed via conda to get promising performance on CPU.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:8]: Note that faiss ...

Comment (hkvision) [reply python/friesian/example/recall/README.md:8]: new line

File: python/friesian/example/recall/README.md [hunk 14]
```diff
@@ -1,6 +1,107 @@
...
+
+```
+conda create -n bigdl python=3.7  # "bigdl" is the conda environment name, you can use any name you like.
+conda activate bigdl
+conda install -c pytorch faiss-cpu 
+pip install --pre --upgrade bigdl-friesian
+```
+
+## Generate data
+You can generate some test data for testing.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:19]: for testing -> to run the example

File: python/friesian/example/recall/README.md [hunk 15]
```diff
@@ -1,6 +1,107 @@
...
+* `index_type`: The faiss index_type: FlatL2 or IVFFlatL2. Default to be FlatL2.
+* `emb_path`: The path to save vector embeddings. Default to be ./emb_vecs.pkl.
+* `dict_path`: The path to save item_dict. Default to be ./item_dict.pkl.
+* `index_save_path`: The path to save faiss index data. Default to be ./index_FlatL2.pkl.
+* `parquet_path`: The path to save vector embeddings with spark, only work when use_spark is True. Default to be ./data.parquet/.
+
+__NOTE:__ 
+The file paths (like '*emb_path*') can be used directly as the corresponding input parameters for the next *search.py*.
+
+## Retrieving items
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:51]: Retrieving -> Search

File: python/friesian/example/recall/README.md [hunk 16]
```diff
@@ -1,6 +1,107 @@
...
+```
+conda create -n bigdl python=3.7  # "bigdl" is the conda environment name, you can use any name you like.
+conda activate bigdl
+conda install -c pytorch faiss-cpu 
+pip install --pre --upgrade bigdl-friesian
+```
+
+## Generate data
+You can generate some test data for testing. 
+The generated data will have a total of **row_nums** items, where each item is a **vec_dim** dimensional vector.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:20]: each item has an id with a vec_dim ...?

File: python/friesian/example/recall/search.py [hunk 17]
```diff
@@ -85,14 +85,9 @@ def do_search(partition):
...
-        executor_cores = args.num_threads
-        executor_memory = "12g"
-        driver_cores = 4
-        driver_memory = "4g"
-        sc = init_orca_context("yarn", cores=executor_cores,
-                               num_nodes=num_executors, memory=executor_memory,
-                               driver_cores=driver_cores, driver_memory=driver_memory,
+        sc = init_orca_context("yarn", cores=args.num_threads,
+                               num_nodes=args.num_tasks, memory=args.memory,
+                               driver_cores=args.num_threads, driver_memory='16g',
```
Comment (hkvision) [inline python/friesian/example/recall/search.py:90]: can we just use the default value of driver cores and driver memory?

File: python/friesian/example/recall/search.py [hunk 18]
```diff
@@ -101,10 +96,10 @@ def search(args):
 
         args.faiss_index_path = args.faiss_index_path.split('/')[-1]
         args.dict_path = args.dict_path.split('/')[-1]
-
+    elif args.cluster_mode == "spark-submit":
+        sc = init_orca_context("spark-submit")
     elif args.cluster_mode == "local":
-        sc = init_orca_context("local", cores=8)
-        num_executors = 4
+        sc = init_orca_context("local")
```
Comment (hkvision) [inline python/friesian/example/recall/search.py:102]: cores=num_threads*num_tasks?

File: python/friesian/example/recall/search.py [hunk 19]
```diff
@@ -135,14 +130,18 @@ def search(args):
 
     parser.add_argument('--num_threads', type=int, default=8,
                         help='Set the environment variable OMP_NUM_THREADS for each faiss task')
-    parser.add_argument('--cluster_mode', type=str, default='yarn',
-                        help='The cluster mode, such as local, yarn')
+    parser.add_argument('--cluster_mode', type=str, default='local',
+                        help='The cluster mode, such as local, yarn or spark-submit')
+    parser.add_argument('--num_tasks', type=int, default=4,
+                        help='The number of nodes to use in the cluster')
```
Comment (hkvision) [inline python/friesian/example/recall/search.py:136]: nodes to use -> faiss tasks to run

File: python/friesian/example/recall/search.py [hunk 20]
```diff
@@ -135,14 +130,18 @@ def search(args):
...
+    parser.add_argument('--memory', type=str, default="50g",
+                        help='The memory limit for each task')
```
Comment (hkvision) [inline python/friesian/example/recall/search.py:138]: The memory allocated for each faiss task

Comment (hkvision) [reply python/friesian/example/recall/search.py:138]: make the default value smaller?

File: python/friesian/example/recall/search.py (hunk 17 above)
Comment (devWangBin) [reply python/friesian/example/recall/search.py:90]: ok

File: python/friesian/example/recall/README.md [hunk 21]
```diff
@@ -1,6 +1,112 @@
...
+conda create -n bigdl python=3.7  # "bigdl" is the conda environment name, you can use any name you like.
+conda activate bigdl
+conda install -c pytorch faiss-cpu 
+pip install --pre --upgrade bigdl-friesian
+```
+
+## Generate data
+You can generate some test data to run the example, which contains:
+- embeddings: item embeddings of the shape [row_nums, vec_dim].
+- item_dict: unique string names of items with the shape [row_nums, ].
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:22]: should describe the key and value?

File: python/friesian/example/recall/README.md [hunk 22]
```diff
@@ -1,6 +1,112 @@
...
+- index data: the faiss index data generated from item embeddings.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:23]: built

File: python/friesian/example/recall/README.md [hunk 23]
```diff
@@ -1,6 +1,112 @@
...
+```
+conda create -n bigdl python=3.7  # "bigdl" is the conda environment name, you can use any name you like.
+conda activate bigdl
+conda install -c pytorch faiss-cpu 
+pip install --pre --upgrade bigdl-friesian
+```
+
+## Generate data
+You can generate some test data to run the example, which contains:
+- embeddings: item embeddings of the shape [row_nums, vec_dim].
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:21]: can we not save embeddings?

File: python/friesian/example/recall/README.md [hunk 24]
```diff
@@ -1,6 +1,112 @@
...
+- item_dict: unique string names of items with the shape [row_nums, ].
+- index data: the faiss index data generated from item embeddings.
+- parquet data: items saved in parquet format, each row contains two values, the id and the embedding.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:24]: items saved in parquet format to do the search, where each row contains item id and its embedding.

File: python/friesian/example/recall/README.md [hunk 25]
```diff
@@ -1,6 +1,112 @@
...
+__Options for generate_test_data:__
+* `row_nums`: The number of vectors to be generated. Default to be 200000.
+* `vec_dim`: The dimension of vector. Default to be 256.
+* `header_len`: The header length of the unique item name. 
+* `verbose`: Print more detail information. Default to be False.
+* `index_type`: The faiss index_type: FlatL2 or IVFFlatL2. Default to be FlatL2.
+* `emb_path`: The path to save vector embeddings. Default to be ./emb_vecs.pkl.
+* `dict_path`: The path to save item_dict. Default to be ./item_dict.pkl.
+* `index_save_path`: The path to save faiss index data. Default to be ./index_FlatL2.pkl.
+* `parquet_path`: The path to save vector embeddings with spark, only work when use_spark is True. Default to be ./data.parquet/.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:49]: modify this, remove with spark related stuff

File: python/friesian/example/recall/README.md [hunk 26]
```diff
@@ -1,6 +1,112 @@
...
+
+__NOTE:__ 
+The file paths (like '*emb_path*') can be used directly as the corresponding input parameters for the next *search.py*.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:52]: can -> will
the next search.py -> search.py below

File: python/friesian/example/recall/README.md [hunk 27]
```diff
@@ -1,6 +1,112 @@
...
+
+## Search items
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:54]: Add simple introduction of this

File: python/friesian/example/recall/README.md [hunk 28]
```diff
@@ -1,6 +1,112 @@
...
+* Spark local, example command:
+```bash
+python search.py \
+    --num_threads 8 \
+    --cluster_mode local \
+    --num_tasks 4 \
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:60]: make the values smaller for local, 2 threads and 2 tasks?

File: python/friesian/example/recall/README.md [hunk 29]
```diff
@@ -1,6 +1,112 @@
...
+    --dict_path /path/to/the/folder/of/item_dict \
+    --faiss_index_path /path/to/the/folder/of/faiss/index/data \
+    --parquet_path /path/to/the/folder/of/vector/embeddings/in/parquet \
+    --parquet_output_path /path/to/the/folder/to/save/retrieval/results 
 ```
+
+__Options for search:__
+* `num_threads`: Set the environment variable OMP_NUM_THREADS for each faiss task. Default to be 8.
+* `cluster_mode`: The cluster mode, one of local, spark-submit or yarn. Default to be local.
+* `num_tasks`: The number of nodes to use in the cluster. Default to be 4.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:101]: The number of faiss tasks to run in the cluster

File: python/friesian/example/recall/README.md [hunk 30]
```diff
@@ -1,6 +1,112 @@
...
+* `memory`: The amount of memory to allocate on each task. Default to be 12g.
+* `top_k`: The number of items to be searched for each query item. Default to be 100.
+* `batch_size`: The batch size for each faiss task. Default to be 50000.
+* `dict_path`: The path to item_dict.pkl. Default to be ./item_dict.pkl.
+* `faiss_index_path`: The path to faiss index data. Default to be ./index_FlatL2.pkl.
+* `parquet_path`: The Path to input parquet data (query items). Default to be ./data.parquet.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:107]: Path -> path

File: python/friesian/example/recall/README.md [hunk 31]
```diff
@@ -1,6 +1,112 @@
...
+* `num_threads`: Set the environment variable OMP_NUM_THREADS for each faiss task. Default to be 8.
+* `cluster_mode`: The cluster mode, one of local, spark-submit or yarn. Default to be local.
+* `num_tasks`: The number of nodes to use in the cluster. Default to be 4.
+* `memory`: The amount of memory to allocate on each task. Default to be 12g.
+* `top_k`: The number of items to be searched for each query item. Default to be 100.
+* `batch_size`: The batch size for each faiss task. Default to be 50000.
+* `dict_path`: The path to item_dict.pkl. Default to be ./item_dict.pkl.
+* `faiss_index_path`: The path to faiss index data. Default to be ./index_FlatL2.pkl.
+* `parquet_path`: The Path to input parquet data (query items). Default to be ./data.parquet. 
+* `parquet_output_path`: The path to save output parquet date (search results). Default to be ./similarity_search_L2.parquet.
```
Comment (hkvision) [inline python/friesian/example/recall/README.md:108]: date -> data

Comment (hkvision) [reply python/friesian/example/recall/README.md:108]: also modify the typo in the code

File: python/friesian/example/recall/README.md (hunk 26 above)
Comment (hkvision) [reply python/friesian/example/recall/README.md:52]: change like emb_path to list the three paths

**Review by hkvision — APPROVED** [review]

**Pull request merged**